### ISSUE TITLE
[PATCH] 1. Convert tabs to spaces in perldsc.pod. 2. Start modernising perldsc.pod

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#14433]
Migrated from [rt.perl.org#123636](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=123636) (status was 'resolved')

Searchable as RT123636$

**Comment by p5pRT** [comment]
### From @shlomif

These patches&#8203;:

1\. Convert tabs to spaces in perldsc\.pod\.

2\. Start modernising perldsc\.pod\.

They also can be found here&#8203;:

https://github.com/shlomif/perl/tree/perldsc-revamp

Regards\,

&nbsp;    Shlomi Fish

P\.S&#8203;: today I discovered that sometimes "make test\_porting" will fail in a
blead clone unless you do "git fetch origin \-\-tags"\. So make sure to fetch the
tags often\.

\-\- 


---
Shlomi Fish       http&#8203;://www\.shlomifish\.org/
Rethinking CPAN \- http&#8203;://shlom\.in/rethinking\-cpan

COBOL is the old Java\.
&nbsp;    — http&#8203;://is\.gd/ClKAz5

Please reply to list if it's a mailing list post \- http&#8203;://shlom\.in/reply \.

**Comment by p5pRT** [comment]
### From @shlomif

<details><summary>0001-perldsc.pod-Convert-tabs-in-indentation-to-spaces.patch</summary>

```diff
From 4dc67d10d2c6f05dda9b208dc8594e72a9624704 Mon Sep 17 00:00:00 2001
From: Shlomi Fish <shlomif@shlomifish.org>
Date: Tue, 20 Jan 2015 16:15:49 +0200
Subject: [PATCH 1/2] [perldsc.pod] Convert tabs in indentation to spaces.

This is so it will display correctly with every tab-step/tab-size.
---
 pod/perldsc.pod | 130 ++++++++++++++++++++++++++++----------------------------
 1 file changed, 65 insertions(+), 65 deletions(-)

diff --git a/pod/perldsc.pod b/pod/perldsc.pod
index c5f53d8..6da0477 100644
--- a/pod/perldsc.pod
+++ b/pod/perldsc.pod
@@ -9,12 +9,12 @@ Perl lets us have complex data structures.  You can write something like
 this and all of a sudden, you'd have an array with three dimensions!
 
     for $x (1 .. 10) {
-	for $y (1 .. 10) {
-	    for $z (1 .. 10) {
-		$AoA[$x][$y][$z] =
-		    $x ** $y + $z;
-	    }
-	}
+        for $y (1 .. 10) {
+            for $z (1 .. 10) {
+                $AoA[$x][$y][$z] =
+                    $x ** $y + $z;
+            }
+        }
     }
 
 Alas, however simple this may appear, underneath it's a much more
@@ -85,10 +85,10 @@ level.  It's just that you can I<use> it as though it were a
 two-dimensional one.  This is actually the way almost all C
 multidimensional arrays work as well.
 
-    $array[7][12]			# array of arrays
-    $array[7]{string}			# array of hashes
-    $hash{string}[7]			# hash of arrays
-    $hash{string}{'another string'}	# hash of hashes
+    $array[7][12]                       # array of arrays
+    $array[7]{string}                   # array of hashes
+    $hash{string}[7]                    # hash of arrays
+    $hash{string}{'another string'}     # hash of hashes
 
 Now, because the top level contains only references, if you try to print
 out your array in with a simple print() function, you'll get something
@@ -116,8 +116,8 @@ repeatedly.  Here's the case where you just get the count instead
 of a nested array:
 
     for $i (1..10) {
-	@array = somefunc($i);
-	$AoA[$i] = @array;	# WRONG!
+        @array = somefunc($i);
+        $AoA[$i] = @array;      # WRONG!
     }
 
 That's just the simple case of assigning an array to a scalar and getting
@@ -125,16 +125,16 @@ its element count.  If that's what you really and truly want, then you
 might do well to consider being a tad more explicit about it, like this:
 
     for $i (1..10) {
-	@array = somefunc($i);
-	$counts[$i] = scalar @array;
+        @array = somefunc($i);
+        $counts[$i] = scalar @array;
     }
 
 Here's the case of taking a reference to the same memory location
 again and again:
 
     for $i (1..10) {
-	@array = somefunc($i);
-	$AoA[$i] = \@array;	# WRONG!
+        @array = somefunc($i);
+        $AoA[$i] = \@array;     # WRONG!
     }
 
 So, what's the big problem with that?  It looks right, doesn't it?
@@ -148,12 +148,12 @@ the following C program:
 
     #include <pwd.h>
     main() {
-	struct passwd *getpwnam(), *rp, *dp;
-	rp = getpwnam("root");
-	dp = getpwnam("daemon");
+        struct passwd *getpwnam(), *rp, *dp;
+        rp = getpwnam("root");
+        dp = getpwnam("daemon");
 
-	printf("daemon name is %s\nroot name is %s\n",
-		dp->pw_name, rp->pw_name);
+        printf("daemon name is %s\nroot name is %s\n",
+                dp->pw_name, rp->pw_name);
     }
 
 Which will print
@@ -169,8 +169,8 @@ broken code fragments:
 X<[]> X<{}>
 
     for $i (1..10) {
-	@array = somefunc($i);
-	$AoA[$i] = [ @array ];
+        @array = somefunc($i);
+        $AoA[$i] = [ @array ];
     }
 
 The square brackets make a reference to a new array with a I<copy>
@@ -181,8 +181,8 @@ Note that this will produce something similar, but it's
 much harder to read:
 
     for $i (1..10) {
-	@array = 0 .. $i;
-	@{$AoA[$i]} = @array;
+        @array = 0 .. $i;
+        @{$AoA[$i]} = @array;
     }
 
 Is it the same?  Well, maybe so--and maybe not.  The subtle difference
@@ -235,9 +235,9 @@ do the right thing behind the scenes.
 
 In summary:
 
-    $AoA[$i] = [ @array ];	# usually best
-    $AoA[$i] = \@array;		# perilous; just how my() was that array?
-    @{ $AoA[$i] } = @array;	# way too tricky for most programmers
+    $AoA[$i] = [ @array ];      # usually best
+    $AoA[$i] = \@array;         # perilous; just how my() was that array?
+    @{ $AoA[$i] } = @array;     # way too tricky for most programmers
 
 
 =head1 CAVEAT ON PRECEDENCE
@@ -247,8 +247,8 @@ Speaking of things like C<@{$AoA[$i]}>, the following are actually the
 same thing:
 X<< -> >>
 
-    $aref->[2][2]	# clear
-    $$aref[2][2]	# confusing
+    $aref->[2][2]       # clear
+    $$aref[2][2]        # confusing
 
 That's because Perl's precedence rules on its five prefix dereferencers
 (which look like someone swearing: C<$ @ * % &>) make them bind more
@@ -279,9 +279,9 @@ also disallow accidental "symbolic dereferencing".  Therefore if you'd done
 this:
 
     my $aref = [
-	[ "fred", "barney", "pebbles", "bambam", "dino", ],
-	[ "homer", "bart", "marge", "maggie", ],
-	[ "george", "jane", "elroy", "judy", ],
+        [ "fred", "barney", "pebbles", "bambam", "dino", ],
+        [ "homer", "bart", "marge", "maggie", ],
+        [ "george", "jane", "elroy", "judy", ],
     ];
 
     print $aref[2][2];
@@ -304,21 +304,21 @@ For example, given the assignment to $AoA above, here's the debugger output:
     DB<1> x $AoA
     $AoA = ARRAY(0x13b5a0)
        0  ARRAY(0x1f0a24)
-	  0  'fred'
-	  1  'barney'
-	  2  'pebbles'
-	  3  'bambam'
-	  4  'dino'
+          0  'fred'
+          1  'barney'
+          2  'pebbles'
+          3  'bambam'
+          4  'dino'
        1  ARRAY(0x13b558)
-	  0  'homer'
-	  1  'bart'
-	  2  'marge'
-	  3  'maggie'
+          0  'homer'
+          1  'bart'
+          2  'marge'
+          3  'maggie'
        2  ARRAY(0x13b540)
-	  0  'george'
-	  1  'jane'
-	  2  'elroy'
-	  3  'judy'
+          0  'george'
+          1  'jane'
+          2  'elroy'
+          3  'judy'
 
 =head1 CODE EXAMPLES
 
@@ -454,10 +454,10 @@ X<hash of arrays> X<HoA>
 
  # print the whole thing sorted by number of members and name
  foreach $family ( sort {
-			    @{$HoA{$b}} <=> @{$HoA{$a}}
-					||
-				    $a cmp $b
-	    } keys %HoA )
+                            @{$HoA{$b}} <=> @{$HoA{$a}}
+                                        ||
+                                    $a cmp $b
+            } keys %HoA )
  {
      print "$family: ", join(", ", sort @{ $HoA{$family} }), "\n";
  }
@@ -560,19 +560,19 @@ X<hash of hashes> X<HoH>
 
  %HoH = (
         flintstones => {
-		lead      => "fred",
-		pal       => "barney",
+                lead      => "fred",
+                pal       => "barney",
         },
         jetsons     => {
-		lead      => "george",
-		wife      => "jane",
-		"his boy" => "elroy",
+                lead      => "george",
+                wife      => "jane",
+                "his boy" => "elroy",
         },
         simpsons    => {
-		lead      => "homer",
-		wife      => "marge",
-		kid       => "bart",
-	},
+                lead      => "homer",
+                wife      => "marge",
+                kid       => "bart",
+        },
  );
 
 =head2 Generation of a HASH OF HASHES
@@ -681,12 +681,12 @@ Here's a sample showing how to create and use a record whose fields are of
 many different sorts:
 
      $rec = {
-	 TEXT      => $string,
-	 SEQUENCE  => [ @old_values ],
-	 LOOKUP    => { %some_table },
-	 THATCODE  => \&some_function,
-	 THISCODE  => sub { $_[0] ** $_[1] },
-	 HANDLE    => \*STDOUT,
+         TEXT      => $string,
+         SEQUENCE  => [ @old_values ],
+         LOOKUP    => { %some_table },
+         THATCODE  => \&some_function,
+         THISCODE  => sub { $_[0] ** $_[1] },
+         HANDLE    => \*STDOUT,
      };
 
      print $rec->{TEXT};
-- 
2.2.2


```
</details>

**Comment by p5pRT** [comment]
### From @shlomif

<details><summary>0002-Modernise-remove-an-anachronistic-comment.patch</summary>

```diff
From 6ae954cf1fdf6ea1c903239ecd9d2b4425c31ed4 Mon Sep 17 00:00:00 2001
From: Shlomi Fish <shlomif@shlomifish.org>
Date: Tue, 20 Jan 2015 16:33:29 +0200
Subject: [PATCH 2/2] Modernise & remove an anachronistic comment.

---
 pod/perldsc.pod | 39 ++++++++++++++++++---------------------
 1 file changed, 18 insertions(+), 21 deletions(-)

diff --git a/pod/perldsc.pod b/pod/perldsc.pod
index 6da0477..8a18455 100644
--- a/pod/perldsc.pod
+++ b/pod/perldsc.pod
@@ -8,9 +8,9 @@ perldsc - Perl Data Structures Cookbook
 Perl lets us have complex data structures.  You can write something like
 this and all of a sudden, you'd have an array with three dimensions!
 
-    for $x (1 .. 10) {
-        for $y (1 .. 10) {
-            for $z (1 .. 10) {
+    for my $x (1 .. 10) {
+        for my $y (1 .. 10) {
+            for my $z (1 .. 10) {
                 $AoA[$x][$y][$z] =
                     $x ** $y + $z;
             }
@@ -94,7 +94,7 @@ Now, because the top level contains only references, if you try to print
 out your array in with a simple print() function, you'll get something
 that doesn't look very nice, like this:
 
-    @AoA = ( [2, 3], [4, 5, 7], [0] );
+    my @AoA = ( [2, 3], [4, 5, 7], [0] );
     print $AoA[1][2];
   7
     print @AoA;
@@ -115,8 +115,8 @@ elements or else taking a reference to the same memory location
 repeatedly.  Here's the case where you just get the count instead
 of a nested array:
 
-    for $i (1..10) {
-        @array = somefunc($i);
+    for my $i (1..10) {
+        my @array = somefunc($i);
         $AoA[$i] = @array;      # WRONG!
     }
 
@@ -124,15 +124,17 @@ That's just the simple case of assigning an array to a scalar and getting
 its element count.  If that's what you really and truly want, then you
 might do well to consider being a tad more explicit about it, like this:
 
-    for $i (1..10) {
-        @array = somefunc($i);
+    for my $i (1..10) {
+        my @array = somefunc($i);
         $counts[$i] = scalar @array;
     }
 
 Here's the case of taking a reference to the same memory location
 again and again:
 
-    for $i (1..10) {
+    # Either without strict or having an outer-scope my @array; declaration.
+
+    for my $i (1..10) {
         @array = somefunc($i);
         $AoA[$i] = \@array;     # WRONG!
     }
@@ -168,7 +170,9 @@ hash constructor C<{}> instead.   Here's the right way to do the preceding
 broken code fragments:
 X<[]> X<{}>
 
-    for $i (1..10) {
+    # Either without strict or having an outer-scope my @array; declaration.
+
+    for my $i (1..10) {
         @array = somefunc($i);
         $AoA[$i] = [ @array ];
     }
@@ -180,7 +184,8 @@ you want.
 Note that this will produce something similar, but it's
 much harder to read:
 
-    for $i (1..10) {
+    # Either without strict or having an outer-scope my @array; declaration.
+    for my $i (1..10) {
         @array = 0 .. $i;
         @{$AoA[$i]} = @array;
     }
@@ -214,7 +219,7 @@ efficient.
 Surprisingly, the following dangerous-looking construct will
 actually work out fine:
 
-    for $i (1..10) {
+    for my $i (1..10) {
         my @array = somefunc($i);
         $AoA[$i] = \@array;
     }
@@ -223,15 +228,7 @@ That's because my() is more of a run-time statement than it is a
 compile-time declaration I<per se>.  This means that the my() variable is
 remade afresh each time through the loop.  So even though it I<looks> as
 though you stored the same variable reference each time, you actually did
-not!  This is a subtle distinction that can produce more efficient code at
-the risk of misleading all but the most experienced of programmers.  So I
-usually advise against teaching it to beginners.  In fact, except for
-passing arguments to functions, I seldom like to see the gimme-a-reference
-operator (backslash) used much at all in code.  Instead, I advise
-beginners that they (and most of the rest of us) should try to use the
-much more easily understood constructors C<[]> and C<{}> instead of
-relying upon lexical (or dynamic) scoping and hidden reference-counting to
-do the right thing behind the scenes.
+not!
 
 In summary:
 
-- 
2.2.2


```
</details>

**Comment by p5pRT** [comment]
### From @jkeenan

On Tue Jan 20 06&#8203;:41&#8203;:31 2015\, shlomif@&#8203;shlomifish\.org wrote&#8203;:
> These patches&#8203;:
> 
> 1\. Convert tabs to spaces in perldsc\.pod\.
> 
> 2\. Start modernising perldsc\.pod\.
> 
> They also can be found here&#8203;:
> 
> https://github.com/shlomif/perl/tree/perldsc-revamp
> 
> Regards\,
> 
> Shlomi Fish
> 

I'll look over these patches in two days and apply them unless someone else beats me to it or objects\.


> P\.S&#8203;: today I discovered that sometimes "make test\_porting" will fail
> in a
> blead clone unless you do "git fetch origin \-\-tags"\. So make sure to
> fetch the
> tags often\.

Can you provide an example?

Thank you very much\.


\-\- 
James E Keenan \(jkeenan@&#8203;cpan\.org\)

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @rjbs

\* James E Keenan via RT \<perlbug\-followup@&#8203;perl\.org> \[2015\-01\-20T21&#8203;:53&#8203;:59\]
> I'll look over these patches in two days and apply them unless someone else
> beats me to it or objects\.

I commented on one commit at gh\.  A large section of prose is removed with only
the indication that it is "anachronistic\."  There is no explanation as to why\.

It doesn't seem so to me\.  I did not review further than that yet\.

\-\- 
rjbs

**Comment by p5pRT** [comment]
### From @jkeenan

On Tue Jan 20 19&#8203;:46&#8203;:31 2015\, perl\.p5p@&#8203;rjbs\.manxome\.org wrote&#8203;:
> \* James E Keenan via RT \<perlbug\-followup@&#8203;perl\.org> \[2015\-01\-
> 20T21&#8203;:53&#8203;:59\]
> > I'll look over these patches in two days and apply them unless
> > someone else
> > beats me to it or objects\.
> 
> I commented on one commit at gh\.  A large section of prose is removed
> with only
> the indication that it is "anachronistic\."  There is no explanation as
> to why\.
> 
> It doesn't seem so to me\.  I did not review further than that yet\.

I applied the first patch to blead in commit 6a40a726ac1d6a8f37527d834a781fc5c38e12f0\.

I won't take any action on the second until someone comments in RT\.

Thank you very much\.

\-\- 
James E Keenan \(jkeenan@&#8203;cpan\.org\)

**Comment by p5pRT** [comment]
### From @shlomif

Hi all\,

here is a modified patch without the controversial change of removing the text\.

\-\- Shlomi Fish

On Tue Jan 20 06&#8203;:41&#8203;:31 2015\, shlomif@&#8203;shlomifish\.org wrote&#8203;:
> These patches&#8203;:
> 
> 1\. Convert tabs to spaces in perldsc\.pod\.
> 
> 2\. Start modernising perldsc\.pod\.
> 
> They also can be found here&#8203;:
> 
> https://github.com/shlomif/perl/tree/perldsc-revamp
> 
> Regards\,
> 
> Shlomi Fish
> 
> P\.S&#8203;: today I discovered that sometimes "make test\_porting" will fail
> in a
> blead clone unless you do "git fetch origin \-\-tags"\. So make sure to
> fetch the
> tags often\.

**Comment by p5pRT** [comment]
### From @shlomif

<details><summary>0001-Modernize-the-code-my-declarations-and-comments.patch</summary>

```diff
From 241ddcd6df20c4d949acfd3bab3cdde669eaf8d7 Mon Sep 17 00:00:00 2001
From: Shlomi Fish <shlomif@shlomifish.org>
Date: Mon, 16 Feb 2015 16:46:29 +0200
Subject: [PATCH] Modernize the code - "my" declarations and comments.

---
 pod/perldsc.pod | 29 +++++++++++++++++------------
 1 file changed, 17 insertions(+), 12 deletions(-)

diff --git a/pod/perldsc.pod b/pod/perldsc.pod
index 6da0477..e2c4eae 100644
--- a/pod/perldsc.pod
+++ b/pod/perldsc.pod
@@ -8,9 +8,9 @@ perldsc - Perl Data Structures Cookbook
 Perl lets us have complex data structures.  You can write something like
 this and all of a sudden, you'd have an array with three dimensions!
 
-    for $x (1 .. 10) {
-        for $y (1 .. 10) {
-            for $z (1 .. 10) {
+    for my $x (1 .. 10) {
+        for my $y (1 .. 10) {
+            for my $z (1 .. 10) {
                 $AoA[$x][$y][$z] =
                     $x ** $y + $z;
             }
@@ -94,7 +94,7 @@ Now, because the top level contains only references, if you try to print
 out your array in with a simple print() function, you'll get something
 that doesn't look very nice, like this:
 
-    @AoA = ( [2, 3], [4, 5, 7], [0] );
+    my @AoA = ( [2, 3], [4, 5, 7], [0] );
     print $AoA[1][2];
   7
     print @AoA;
@@ -115,8 +115,8 @@ elements or else taking a reference to the same memory location
 repeatedly.  Here's the case where you just get the count instead
 of a nested array:
 
-    for $i (1..10) {
-        @array = somefunc($i);
+    for my $i (1..10) {
+        my @array = somefunc($i);
         $AoA[$i] = @array;      # WRONG!
     }
 
@@ -124,15 +124,17 @@ That's just the simple case of assigning an array to a scalar and getting
 its element count.  If that's what you really and truly want, then you
 might do well to consider being a tad more explicit about it, like this:
 
-    for $i (1..10) {
-        @array = somefunc($i);
+    for my $i (1..10) {
+        my @array = somefunc($i);
         $counts[$i] = scalar @array;
     }
 
 Here's the case of taking a reference to the same memory location
 again and again:
 
-    for $i (1..10) {
+    # Either without strict or having an outer-scope my @array; declaration.
+
+    for my $i (1..10) {
         @array = somefunc($i);
         $AoA[$i] = \@array;     # WRONG!
     }
@@ -168,7 +170,9 @@ hash constructor C<{}> instead.   Here's the right way to do the preceding
 broken code fragments:
 X<[]> X<{}>
 
-    for $i (1..10) {
+    # Either without strict or having an outer-scope my @array; declaration.
+
+    for my $i (1..10) {
         @array = somefunc($i);
         $AoA[$i] = [ @array ];
     }
@@ -180,7 +184,8 @@ you want.
 Note that this will produce something similar, but it's
 much harder to read:
 
-    for $i (1..10) {
+    # Either without strict or having an outer-scope my @array; declaration.
+    for my $i (1..10) {
         @array = 0 .. $i;
         @{$AoA[$i]} = @array;
     }
@@ -214,7 +219,7 @@ efficient.
 Surprisingly, the following dangerous-looking construct will
 actually work out fine:
 
-    for $i (1..10) {
+    for my $i (1..10) {
         my @array = somefunc($i);
         $AoA[$i] = \@array;
     }
-- 
2.3.0


```
</details>

**Comment by p5pRT** [comment]
### From @jkeenan

Most recent patch applied in commit 5939083a782eb3d8f49ef8e3a7802a90daadc5f6\.

Marking ticket Resolved\.

Thank you very much\.

\-\- 
James E Keenan \(jkeenan@&#8203;cpan\.org\)

**Comment by p5pRT** [comment]
@jkeenan - Status changed from 'open' to 'resolved'
